driver/razermouse_driver.c

```diff
@@ -1331,6 +1331,114 @@ static ssize_t razer_attr_read_mouse_dpi(struct device *dev, struct device_attri
     return sprintf(buf, "%u:%u\n", dpi_x, dpi_y);
 }
 
+/**
+ * Write device file "dpi_stages"
+ *
+ * Sets the mouse DPI stages to the array of unsigned shorts written to this
+ * file.
+ * The number of DPI stages is hard limited by RAZER_MOUSE_MAX_DPI_STAGES.
+ *
+ * Each DPI stage is described by 4 bytes:
+ *   - 2 bytes (unsigned short) for x-axis DPI
+ *   - 2 bytes (unsigned short) for y-axis DPI
+ *
+ * buf is expected to contain n*4 bytes where n is the number of DPI stages.
+ * If count is not a multiple of 4 then the residual bytes will be ignored.
+ *
+ * Example: let's say you want to set the following DPI stages:
+ *  (800, 800), (1800, 1800), (3600, 3200)  // (DPI X, DPI Y)
+ *
+ * You have to write to this file 6 unsigned shorts (big endian) = 12 bytes:
+ *   DPIs:         800 | 800 | 1800 | 1800 | 3600 | 3200
+ *   Bytes (hex): 03 20 03 02 07 08  07 08  0e 10  0c 80
+ */
+static ssize_t razer_attr_write_mouse_dpi_stages(struct device *dev, struct device_attribute *attr, const char *buf, size_t count)
+{
+    struct razer_mouse_device *device = dev_get_drvdata(dev);
+    struct razer_report report = {0};
+    unsigned short dpi[2 * RAZER_MOUSE_MAX_DPI_STAGES] = {0};
+    unsigned char stages_count = 0;
+    size_t remaining = count;
+
+    while (stages_count < RAZER_MOUSE_MAX_DPI_STAGES && remaining >= 4) {
+        // DPI X
+        dpi[stages_count * 2]     = (buf[0] << 8) | (buf[1] & 0xFF);
+
+        // DPI Y
+        dpi[stages_count * 2 + 1] = (buf[2] << 8) | (buf[3] & 0xFF);
+
+        stages_count += 1;
+        buf += 4;
+        remaining -= 4;
+    }
+
+    report = razer_chroma_misc_set_dpi_stages(VARSTORE, stages_count, dpi);
+
+    razer_send_payload(device->usb_dev, &report);
+
+    // Always return count, otherwise some programs can enter an infinite loop.
+    // Example:
+    // Program writes 7 bytes to dpi_stages. 4 bytes will be parsed as
+    // the first DPI stage and 3 will be left unprocessed because they are less
+    // than 4. The program will try to write the 3 bytes again but this
+    // function will always return 0, throwing the program into a loop.
+    return count;
+}
+
+/**
+ * Read device file "dpi_stages"
+ *
+ * Writes comma-separated list of "dpi_x:dpi_y" (one for each DPI stage) to
+ * buf.
```

Comment:
This doc is outdated now, you can probably either copy the doc from the setter or just say something like "See write method above" - I believe the latter is better

Reply:
Done in 8d54135, sorry I forgot about that.